python-checks/src/main/java/org/sonar/python/checks/UselessParenthesisCheck.java

```diff
@@ -19,93 +19,142 @@
-    subscribeTo(
-        PythonGrammar.TEST,
-        PythonGrammar.EXPR,
-        PythonGrammar.NOT_TEST
-    );
+  public void visitFile(AstNode astNode) {
+    leftParenthesisDataStack.clear();
+//    pairsByLevel.clear();
+    previousTokenWasLeftParenthesis = false;
+    previousTokenWasRightParenthesis = false;
   }
 
   @Override
-  public void visitNode(AstNode node) {
-    if (node.is(PythonGrammar.NOT_TEST)) {
-      visitNotTest(node);
-    } else if (node.getNumberOfChildren() == 1 && node.getFirstChild().is(PythonGrammar.ATOM)) {
-
-      AstNode atom = node.getFirstChild();
-      AstSelect selectedParent = atom.select().parent().parent().parent();
-      if (selectedParent.size() == 1) {
-        AstNode parent = selectedParent.get(0);
-        if (isKeywordException(parent)) {
-          checkAtom(atom, true);
-          return;
-        }
+  public void visitToken(Token token) {
+    if (token.getType() == LPARENTHESIS) {
+      if (previousTokenWasLeftParenthesis) {
+        leftParenthesisDataStack.peek().setContiguousLeftParenthesis(token);
       }
+      leftParenthesisDataStack.push(new LeftParenthensisData(token));
+      previousTokenWasLeftParenthesis = true;
+      previousTokenWasRightParenthesis = false;
+
+    } else if (token.getType() == RPARENTHESIS) {
+      checkRedundantParentheses(token);
+//      cleanUpPrevious();
+      previousTokenWasLeftParenthesis = false;
+      previousTokenWasRightParenthesis = true;
+      last = leftParenthesisDataStack.pop();
 
-      checkAtom(atom, false);
+    } else {
+      previousTokenWasLeftParenthesis = false;
+      previousTokenWasRightParenthesis = false;
     }
   }
 
-  private static boolean isKeywordException(AstNode parent) {
-    if ((parent.is(PythonGrammar.RETURN_STMT) || parent.is(PythonGrammar.YIELD_EXPR)) && parent.getFirstChild(PythonGrammar.TESTLIST).getNumberOfChildren() == 1) {
-      return true;
+  /**
+   * Method to prevent a FP on cases such as <code>y = ((x1) * (x2))</code>.
+  private void cleanUpPrevious() {
+    LeftParenthensisData previous = leftParenthesisDataStack.peek();
+    if (previous != null) {
+      previous.setFollowedByLeftParenthesis(false);
     }
-    return parent.is(PythonGrammar.FOR_STMT) && parent.getFirstChild(PythonGrammar.EXPRLIST).getChildren(PythonGrammar.EXPR).size() == 1;
   }
+   */
 
-  private void checkAtom(AstNode atom, boolean ignoreTestNumber) {
-    if (violationCondition(atom, ignoreTestNumber)) {
-      addIssue(atom, "Remove those useless parentheses");
+  private void checkRedundantParentheses(Token rightParenthesis) {
+    LeftParenthensisData current = leftParenthesisDataStack.peek();
+//    LeftParenthensisData enclosing = leftParenthesisDataStack.get(leftParenthesisDataStack.size() - 2);
+    
+    if (previousTokenWasRightParenthesis && current.getContiguousLeftParenthesis() == last.getLeftParenthesis()) {  
+//      && enclosing.getContiguousLeftParenthesis() == current.getLeftParenthesis()) {
```

Comment:
![MAJOR](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/severity-major.png 'Severity: MAJOR') This block of commented-out lines of code should be removed. [![rule](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/rule.png)](https://next.sonarqube.com/sonarqube/coding_rules#rule_key=squid%3ACommentedOutCodeLine)